pkg/apis/auditregistration/types.go

```diff
@@ -59,72 +59,98 @@ const (
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// AuditSink represents a cluster level sink for audit data
+// AuditSink represents a cluster level audit sink
 type AuditSink struct {
 	metav1.TypeMeta
-
 	// +optional
 	metav1.ObjectMeta
 
-	// Spec defines the audit sink spec
+	// `spec` defines the `auditSink` spec
 	Spec AuditSinkSpec
 }
 
-// AuditSinkSpec is the spec for the audit sink object
+// AuditSinkSpec holds the spec for the audit sink
 type AuditSinkSpec struct {
-	// Policy defines the policy for selecting which events should be sent to the backend
+	// `policy` defines the policy for selecting which events should be sent to the webhook
 	// required
 	Policy Policy
 
-	// Webhook to send events
+	// `webhook` to send events
 	// required
 	Webhook Webhook
 }
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// AuditSinkList is a list of a	audit sink items.
+// AuditSinkList is a list of AuditSink items.
 type AuditSinkList struct {
 	metav1.TypeMeta
-
 	// +optional
 	metav1.ListMeta
 
-	// List of audit configurations.
+	// `items` is a list of `auditSink` configurations.
 	Items []AuditSink
 }
 
 // Policy defines the configuration of how audit events are logged
 type Policy struct {
-	// The Level that all requests are recorded at.
+	// `level` is the base level that all requests are recorded at.
+	// available options: None, Metadata, Request, RequestResponse
+	// required
+	Level Level
+
+	// `stages` is a list of stages for which events are created.
+	// If no stages are given nothing will be logged
+	// +optional
+	Stages []Stage
+
+	// `rules` define how `auditClass` objects should be handled.
+	// A request may fall under multiple `auditClass` objects.
+	// Rules are evaluated in order (first matching wins).
+	// Rules take precedence over the the parent `level` & `stages`.
+	// Unmatched requests use the parent `level` rule.
+	// +optional
+	Rules []PolicyRule
+}
+
+// PolicyRule defines what level and stages are recorded for an `auditClass`.
+type PolicyRule struct {
+	// `withAuditClass` is the name of the `auditClass` object. This rule matches requests that are
+	// classified with this `auditClass`
+	// required
+	WithAuditClass string
+
+	// `level` that all requests this rule applies to are recorded at.
 	// available options: None, Metadata, Request, RequestResponse
+	// This will override the parent policy Level.
 	// required
 	Level Level
 
-	// Stages is a list of stages for which events are created.
+	// `stages` is a list of stages for which events are created.
+	// If no stages are given nothing will be logged.
```

Comment:
Related to my comment wrt Stages n Policy level. If no stages have been specified here, nothing will be logged even if the Policy specifies stages of its own? Can we make the correlation more explicit?

Reply:
Also does this works as a union with the Policy stages as in the audit API or it completely overrides it? This should also be made explicit.